File: transcripts/turns.py
"""Task #37 — speaker-turn coalescing (a display PROJECTION, never a raw mutation).

The pipeline emits many small per-span segments (diart streams per-finalized-span;
DiariZen also spans one person's speech). Rendering one row per span makes a single
speaker read as N fragments. `group_into_turns` merges **consecutive same-speaker
spans into one turn**, closing a turn only on a real speaker change.

Core invariant: `raw_diarization` stays span-level and immutable — a turn WRAPS its
spans (keeps their `start`/`end`/`text` under `turn["spans"]`) so per-segment clips
(#3), the v2 editor (#9), and click-to-seek (#41) still operate on spans. This is a
deterministic view; re-running it after a tag / #13-heal (names change → grouping
changes) is free.

Shared with the frontend: `frontend/src/lib/turns.ts` mirrors this exactly (the live
page coalesces the streaming spans client-side). Keep the two in lockstep.
"""
from __future__ import annotations

from typing import Any

#: Same-speaker gap (seconds) beyond which we insert a PARAGRAPH BREAK inside the
#: turn (still one turn — never a split). Configurable; a long monologue pause reads
#: better broken up. Tune here + in turns.ts together.
PARAGRAPH_GAP_SEC = 10.0


def _speaker_key(seg: dict) -> tuple:
    """The STABLE merge key for a span. Prefer the resolved identity so the same
    person merges even across different diarizer local labels; fall back to the raw
    local label so two DISTINCT unknown speakers (different labels) never merge.

    A `proposed_name` (recognized-but-unconsented, #3) is NOT a merge key — it isn't
    confirmed, so it must not silently glue two speakers together.
    """
    vid = seg.get("voiceprint_id")
    if vid:
        return ("vp", vid)
    name = seg.get("speaker_name")
    if name:
        return ("name", name)
    return ("local", seg.get("speaker"))


#: Fields copied from a turn's first (text-bearing) span onto the turn itself, so the
#: UI renders the speaker header without re-joining resolved_speakers.
_IDENTITY_FIELDS = ("speaker", "speaker_name", "proposed_name", "voiceprint_id", "consented")


def _join_text(prev_text: str, prev_end: Any, seg: dict) -> str:
    """Append a span's text to the turn's running text — a single space between
    spans, or a paragraph break when the same speaker paused longer than
    `PARAGRAPH_GAP_SEC`."""
    text = (seg.get("text") or "").strip()
    if not prev_text:
        return text
    start = seg.get("start")
    gap = None
    if isinstance(start, (int, float)) and isinstance(prev_end, (int, float)):
        gap = start - prev_end
    sep = "\n\n" if (gap is not None and gap > PARAGRAPH_GAP_SEC) else " "
    return f"{prev_text}{sep}{text}"


def group_into_turns(segments: list[dict]) -> list[dict]:
    """Coalesce consecutive same-speaker spans into turns.

    Each turn: ``{speaker, speaker_name, proposed_name, voiceprint_id, consented,
    start (first span), end (last span), text (joined), spans: [<original span
    dicts>]}``. Empty/whitespace spans never OPEN or FLIP a turn — they're absorbed
    into the open turn (span kept, `end` extended) or skipped when nothing is open.
    Input order is preserved (callers pass time-ordered spans).
    """
    turns: list[dict] = []
    open_turn: dict | None = None
    open_key: tuple | None = None

    for seg in segments:
        text = (seg.get("text") or "").strip()
        if not text:
            # Silence / empty span: never opens or flips a turn. Absorb into the
            # current turn (keep the span, extend end) so clips/seek still see it.
            if open_turn is not None:
                open_turn["spans"].append(seg)
                if seg.get("end") is not None:
                    open_turn["end"] = seg.get("end")
            continue

        key = _speaker_key(seg)
        if open_turn is not None and key == open_key:
            open_turn["text"] = _join_text(open_turn["text"], open_turn["_prev_end"], seg)
            open_turn["spans"].append(seg)
            if seg.get("end") is not None:
                open_turn["end"] = seg.get("end")
            open_turn["_prev_end"] = seg.get("end")
            continue

        # Speaker change (or first span) → close the open turn, open a new one.
        if open_turn is not None:
            turns.append(_finalize(open_turn))
        open_turn = {
            **{f: seg.get(f) for f in _IDENTITY_FIELDS},
            "start": seg.get("start"),
            "end": seg.get("end"),
            "text": text,
            "spans": [seg],
            "_prev_end": seg.get("end"),
        }
        open_key = key

    if open_turn is not None:
        turns.append(_finalize(open_turn))
    return turns


def _finalize(turn: dict) -> dict:
    turn.pop("_prev_end", None)  # internal join cursor
    return turn

File: frontend/src/lib/turns.ts
/**
 * Task #37 — speaker-turn coalescing on the client (mirrors `transcripts/turns.py`).
 *
 * Coalesce consecutive same-speaker spans into one turn — used by the live page
 * (over the streaming spans, so the open turn GROWS) and the refine editor (to group
 * v2 segments under one speaker header). The read view uses the server-computed
 * `turns` directly. Keep this in lockstep with the Python.
 *
 * Projection only: a turn WRAPS its spans (`turn.spans`) so clips/edit/seek keep
 * operating on the underlying spans.
 */

/** Same-speaker gap (s) beyond which we insert a paragraph break inside the turn. */
export const PARAGRAPH_GAP_SEC = 10.0;

export type SpeakerKeyable = {
  speaker?: string | null;
  speaker_label?: string | null;
  speaker_name?: string | null;
  voiceprint_id?: string | null;
  proposed_name?: string | null;
};

/**
 * Stable merge key: resolved identity (voiceprint → confirmed name) wins so the same
 * person merges across different diarizer local labels; otherwise the raw local label
 * keeps two distinct unknowns apart. A `proposed_name` (unconsented) is NOT a key.
 */
export function speakerKey(seg: SpeakerKeyable): string {
  if (seg.voiceprint_id) return `vp:${seg.voiceprint_id}`;
  if (seg.speaker_name) return `name:${seg.speaker_name}`;
  return `local:${seg.speaker ?? seg.speaker_label ?? ""}`;
}

/** Consecutive runs of items sharing a key (the primitive both the editor + live use). */
export function groupConsecutive<T>(items: T[], keyOf: (t: T) => string): T[][] {
  const out: T[][] = [];
  let cur: T[] | null = null;
  let curKey: string | null = null;
  for (const it of items) {
    const k = keyOf(it);
    if (cur && k === curKey) {
      cur.push(it);
    } else {
      cur = [it];
      curKey = k;
      out.push(cur);
    }
  }
  return out;
}

type Span = SpeakerKeyable & { text?: string | null; start?: number | null; end?: number | null };

export type Turn<S extends Span> = {
  speaker?: string | null;
  speaker_name?: string | null;
  proposed_name?: string | null;
  voiceprint_id?: string | null;
  consented?: boolean | null;
  start: number | null;
  end: number | null;
  text: string;
  spans: S[];
};

/**
 * Full turn coalescing over time-ordered spans (live page). Empty/whitespace spans
 * never open or flip a turn — they're absorbed into the open turn (kept, `end`
 * extended). A big same-speaker gap inserts a paragraph break (still one turn).
 */
export function groupIntoTurns<S extends Span>(segments: S[]): Turn<S>[] {
  const turns: Turn<S>[] = [];
  let open: (Turn<S> & { _prevEnd: number | null }) | null = null;
  let openKey: string | null = null;

  for (const seg of segments) {
    const text = (seg.text ?? "").trim();
    if (!text) {
      if (open) {
        open.spans.push(seg);
        if (seg.end != null) open.end = seg.end;
      }
      continue;
    }
    const key = speakerKey(seg);
    if (open && key === openKey) {
      const gap =
        seg.start != null && open._prevEnd != null ? seg.start - open._prevEnd : null;
      const sep = gap != null && gap > PARAGRAPH_GAP_SEC ? "\n\n" : " ";
      open.text = open.text ? `${open.text}${sep}${text}` : text;
      open.spans.push(seg);
      if (seg.end != null) open.end = seg.end;
      open._prevEnd = seg.end ?? open._prevEnd;
      continue;
    }
    if (open) turns.push(strip(open));
    open = {
      speaker: seg.speaker ?? seg.speaker_label ?? null,
      speaker_name: seg.speaker_name ?? null,
      proposed_name: seg.proposed_name ?? null,
      voiceprint_id: seg.voiceprint_id ?? null,
      consented: (seg as { consented?: boolean | null }).consented ?? null,
      start: seg.start ?? null,
      end: seg.end ?? null,
      text,
      spans: [seg],
      _prevEnd: seg.end ?? null,
    };
    openKey = key;
  }
  if (open) turns.push(strip(open));
  return turns;
}

function strip<S extends Span>(t: Turn<S> & { _prevEnd: number | null }): Turn<S> {
  const { _prevEnd, ...rest } = t;
  void _prevEnd;
  return rest;
}
